File: Leetcode 3901-4000/3926/3926.py
# Leetcode 3926: Count Valid Word Occurrences
# https://leetcode.com/problems/count-valid-word-occurrences/
# Solved on 24th of May, 2026
class Solution:
    def countWordOccurrences(self, chunks: list[str], queries:  list[str]) -> list[int]:
        """
        Counts the occurrences of specific query words within a string formed by joining multiple chunks.

        :param chunks: A list of strings that are concatenated to form the source text.
        :param queries: A list of strings representing the words to count in the source text.
        :return: A list of integers where each integer is the frequency of the corresponding query word.
        """
        joinedString = "".join(chunks)
        strLen = len(joinedString)
        wordFreq = {}
        currWord = []

        for i in range(strLen):
            char = joinedString[i]
            if 'a' <= char <= 'z':
                currWord.append(char)
            elif char == '-' and i > 0 and i + 1 < strLen and 'a' <= joinedString[i - 1] <= 'z' and 'a' <= joinedString[i + 1] <= 'z':
                currWord.append(char)
            else:
                if currWord:
                    wordStr = "".join(currWord)
                    wordFreq[wordStr] = wordFreq.get(wordStr, 0) + 1
                    currWord = []

        if currWord:
            wordStr = "".join(currWord)
            wordFreq[wordStr] = wordFreq.get(wordStr, 0) + 1

        ansArray = []
        for queryStr in queries:
            ansArray.append(wordFreq.get(queryStr, 0))

        return ansArray

File: Leetcode 3901-4000/3926/3926.java
// Leetcode 3926: Count Valid Word Occurrences
// https://leetcode.com/problems/count-valid-word-occurrences/
// Solved on 24th of May, 2026
import java.util.HashMap;
import java.util.Map;

class Solution {
    /**
     * Counts the occurrences of specific query words within a text formed by concatenating multiple chunks.
     *
     * @param chunks  An array of strings that, when concatenated, form the full text to be analyzed.
     * @param queries An array of strings representing the words to search for in the processed text.
     * @return An array of integers where each element represents the frequency of the corresponding query word.
     */
    public int[] countWordOccurrences(String[] chunks, String[] queries) {
        StringBuilder builder = new StringBuilder();
        for (String chunk : chunks) {
            builder.append(chunk);
        }
        
        char[] characters = builder.toString().toCharArray();
        int length = characters.length;
        Map<String, Integer> wordCount = new HashMap<>();
        StringBuilder wordBuilder = new StringBuilder();
        
        for (int i = 0; i < length; i++) {
            char current = characters[i];
            if (current >= 'a' && current <= 'z') {
                wordBuilder.append(current);
            } else if (current == '-') {
                if (i > 0 && i < length - 1 && 
                    characters[i - 1] >= 'a' && characters[i - 1] <= 'z' && 
                    characters[i + 1] >= 'a' && characters[i + 1] <= 'z') {
                    wordBuilder.append(current);
                } else {
                    if (wordBuilder.length() > 0) {
                        String word = wordBuilder.toString();
                        wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
                        wordBuilder.setLength(0);
                    }
                }
            } else {
                if (wordBuilder.length() > 0) {
                    String word = wordBuilder.toString();
                    wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
                    wordBuilder.setLength(0);
                }
            }
        }
        
        if (wordBuilder.length() > 0) {
            String word = wordBuilder.toString();
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }
        
        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            result[i] = wordCount.getOrDefault(queries[i], 0);
        }
        
        return result;
    }
}